### PR TITLE
Adding a --info flag to output meta information of the model

### PR DESCRIPTION
**Pull request by mandel**
## Release notes

Proposal for issue #808.

Add an option `--info` which displays information about the Stan model. It currently generates a JSON object with a field `inputs`, `parameters`, `transformed parameters`, and `generated quantities` containing a dictionary where each entry corresponds to a variable in respectively the `data`, `parameters`, `transformed parameters`, and `generated quantities` blocks. To each variable is associated an object with two fields:
- `type`: the base type of the variable ("int" or "real").
- `dimensions`: the number of dimensions (0 for a scalar, 1 for a vector or row vector, etc.).

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Review by seantalts — CHANGES_REQUESTED** [review]
Code looks great, thanks! Would you mind adding a test to https://github.com/stan-dev/stanc3/tree/master/test/integration/cli-args for this? Thank you!

**Comment by mandel** [comment]
Thanks, it helped me identify a bug!

File: test/integration/cli-args/info.stan [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  ordered[12] n;
+  positive_ordered[13] o;
+  cov_matrix[14] p;
+  corr_matrix[15] q;
+  cholesky_factor_cov[16] r;
+  cholesky_factor_corr[17] s;
+}
+
+transformed parameters {
+  matrix[14, 14] t = p;
```
Comment (rok-cesnovar) [inline test/integration/cli-args/info.stan:30]:
```suggestion
  matrix[14, 14] t = p;
  {
    real ignored_tp = r;
  }
```

File: test/integration/cli-args/info.stan [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+model {
+
```
Comment (rok-cesnovar) [inline test/integration/cli-args/info.stan:34]:
```suggestion
model {
    real ignored_model = r;
```

File: test/integration/cli-args/info.stan [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+generated quantities {
+  real u = l[0];
```
Comment (rok-cesnovar) [inline test/integration/cli-args/info.stan:38]:
```suggestion
  real u = l[0];
  {
    real ignored_gq = l[1];
  }
```

**Review by rok-cesnovar — COMMENTED** [review]
A few suggestions to expand  the test with properly ignored output variables.

**Comment by mandel** [comment]
Thank @rok-cesnovar you for your suggestions!

**Comment by seantalts** [comment]
I made a couple minor edits in this PR (didn't have permission to just push them upstream to the info branch): https://github.com/deepppl/stanc3/pull/2

We're trying to use `Fmt` when possible, so switched to doing that, and also moved the `exit 0` and `ref` controlling info printing to `stanc.ml` since we're trying to move most of the I/O and interaction with the world into the driver when possible.

**Review by seantalts — APPROVED** [review]
Woo, thank you! This is great and should be really useful for anyone building stuff with or on top of Stan :D

**Pull request merged**